### PR TITLE
[None][chore] Zero FlashInfer split-KV scratch; gate the per-pass reset behind an opt-in flag

### PR DESCRIPTION
**Pull request by brnguyen2**
## Description

The FlashInfer split-KV attention path uses the shared float workspace as run scratch: the attention kernel carves per-tile partial outputs (`tmp_v`) and log-sum-exps (`tmp_s`) out of it and a merge pass combines them. That workspace is allocated with `torch.empty`, reused across differently-shaped plans and CUDA-graph replays, and never cleared. This change removes the merged attention result's latent dependence on whatever an earlier plan — or the allocator — left in that scratch.

This is **defensive hardening, not a fix for a known reproducer.** On current FlashInfer the merge only reads tiles the same pass wrote (the write index and merge range derive from the same per-request chunk count), so the stale-partial hazard the per-pass reset guards against cannot be constructed. The value is in not depending on that scheduler invariant holding for every plan shape and FlashInfer version.

This PR:

* **Zeroes the workspace once at allocation** (always on). Cheap — one memset per engine — and removes any dependence on allocator history for the first pass.
* **Optionally resets each plan's `tmp_v`/`tmp_s` extent before `wrapper.run()`** for the three non-MLA run paths (paged prefill, paged decode, ragged prefill), behind the `TLLM_FLASHINFER_RESET_SPLIT_KV_SCRATCH` env var (default off). Extents are read from the wrapper's live plan info; the caller tells the helper whether it holds a prefill or a decode wrapper, so the layout is selected explicitly rather than guessed from the plan-info length, and non-splitting plans plus other backends' layouts are skipped. When enabled, the plan-info offsets are pinned to a known FlashInfer version and a mismatch fails loudly at metadata build time, so a version bump cannot silently zero the wrong bytes.

**Placement.** A CUDA graph is captured around the run and replayed without re-planning, so a reset at plan time would run once and every replay would attend over what the previous replay left. Before `run()` it is captured into the graph with constant bounds and re-runs on each replay; it is equally correct when a plan is reused across steps, because the extent comes from the wrapper's live plan info.

**Sentinel.** `tmp_v` is zeroed, but zero is the wrong value for `tmp_s`: the online-softmax merge starts from m = −inf, d = 1, so a stale `tmp_s` = 0 contributes `exp2(0 − m)` and attenuates a real partial; −inf gives `exp2(−inf − −inf)` = NaN. A large finite-negative sentinel is neutral on both counts.

**Scope.** MLA wrappers are untouched.

**Cost.** The per-pass reset is a memset of a few MiB per attention layer per pass (order ~100 µs per decode step across ~60 layers on HBM-class bandwidth). Measured as not visible in generation time on an older tree; not yet re-measured on current `main`.

## Test Coverage

* CPU-only unit tests, `tests/unittest/_torch/attention/test_flashinfer_attention.py::TestSplitKVScratchReset`, covering the prefill and decode reset extents, the sentinel's finiteness, that no byte outside the plan's two ranges is touched, that non-splitting and other-backend plans are a no-op, and that the version guard rejects an unpinned FlashInfer. They exercise the helper against a host workspace tensor and a stub wrapper, so no GPU or planned FlashInfer wrapper is needed.
* GPU validation on current `main` is pending. There is no reproducer this change alone fixes.

## Dev Engineer Review

- `flashinfer.py` zero-initializes shared workspace allocation.
- Opt-in reset logic supports explicit prefill and decode layouts.
- Bounds and version checks reduce layout and memory risks.
- The reset option requires FlashInfer `0.6.18`.
- MLA wrappers remain unchanged.
- The public method now requires keyword-only `is_decode`.
- Reset remains disabled by default.

## QA Engineer Review

- `test_flashinfer_attention.py` adds CPU-only coverage for reset extents, sentinel finiteness, untouched and out-of-range ranges, workspace initialization, version validation, and no-op plans.
- Coverage includes prefill and decode paths, non-splitting plans, other-backend plans, and error handling.
- No matching entry exists in the CI or manual-QA test lists.
- GPU validation covered reset tests, the Gemma 3 1B IT smoke test, repeatability, and guard-page diagnostics.
- Verdict: sufficient.

## Per-File QA Perspective

- `tensorrt_llm/_torch/attention/backends/flashinfer.py`: Verify workspace initialization, opt-in resets for ragged prefill, paged prefill, and decode, layout validation, version failures, and unchanged MLA behavior.
- `tests/unittest/_torch/attention/test_flashinfer_attention.py`: Covers reset dimensions, sentinel values, untouched and out-of-range ranges, workspace initialization, no-op paths, and version rejection. No matching CI or manual-QA list entry exists.

### CODE REVIEW

**Comment by brnguyen2** [comment]
GPU validation (fresh build of this branch merged with #18754 and #18756 on top of current main, GB300 / SBSA, PyTorch backend):

- Unit: `tests/unittest/_torch/attention/test_flashinfer_attention.py::TestSplitKVScratchReset` 2/2 passed.
- Smoke: gemma-3-1b-it (sliding window 512) with `attn_backend=FLASHINFER`, CUDA graphs with padding, KV cache manager v2, greedy; 10 mixed prompts of 6 to 5059 tokens, 6 of which cross the window. All 10 completions finite and non-empty; a repeat run produced identical token ids. Same run with the #18754/#18756 diagnostics enabled (`TRTLLM_FI_PAGE_TABLE_CHECK=1`, guard page and fresh-page fill set to NaN): token ids identical to the defaults run, zero page-table fault reports.

Marking ready for review.

File: tensorrt_llm/_torch/attention/backends/flashinfer.py [hunk 1]
```diff
@@ -1471,6 +1507,59 @@ def _refresh_fa2_cuda_graph_plans(self) -> None:
...
+        everything it depends on is fixed for the life of a captured graph.
+        """
+        plan_info = getattr(wrapper, "_plan_info", None)
+        if plan_info is None or self.workspace_buffer is None:
+            return
+        plan_info = list(plan_info)
+        # The FA2 batch-prefill plan returns 15 entries; the batch-decode plan
+        # returns 10 in a different order, so the length picks out the layout.
+        if len(plan_info) != _FI_PREFILL_PLAN_INFO_LEN:
+            return
```
[the remark this answers is not in the document]
Comment (brnguyen2) [reply tensorrt_llm/_torch/attention/backends/flashinfer.py]: Fixed in 572b931f128. Verified against FlashInfer 0.6.18's `scheduler.cuh` (`DecodePlanInfo::ToVector`): the indices here are correct -- `padded_batch_size`/`v_offset`/`s_offset` at entries 0/1/2 and `split_kv` at entry 9. `zero_planned_split_kv_scratch()` now recognizes the 10-entry decode plan and resets its planned `tmp_v`/`tmp_s` extents. Note the decode extent has no `cta_tile_q` factor: the decode plan allocates `tmp_v` as `num_qo_heads * padded_batch_size * head_dim` floats (one query row per request) and `tmp_s` the same without `head_dim`, so the reset uses `padded_batch_size * num_qo_heads` rows. Added `TestSplitKVScratchReset::test_resets_exactly_the_planned_decode_extent` and kept the no-op coverage for non-split prefill/decode plans and unrecognized plan layouts; all three tests pass.

File: tensorrt_llm/_torch/attention/backends/flashinfer.py [hunk 2]
```diff
@@ -1471,6 +1513,73 @@ def _refresh_fa2_cuda_graph_plans(self) -> None:
...
+        plan_info = list(plan_info)
+        # The FA2 batch-prefill plan returns 15 entries; the batch-decode plan
+        # returns 10 in a different order, so the length picks out the layout.
+        # The extents mirror the plans' allocations: prefill sizes tmp_v as
+        # padded_batch_size * cta_tile_q * num_qo_heads * head_dim rows of
+        # float, decode as padded_batch_size * num_qo_heads * head_dim (one
+        # query row per request, no query tiling); tmp_s is the same without
+        # the head_dim factor.
+        if len(plan_info) == _FI_PREFILL_PLAN_INFO_LEN:
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention/backends/flashinfer.py]: Why can we ensure that only FA2 prefill plan info has this length? If there are other plan info use the same length, we will crash it. Ia there safer way to detect plan info type?

File: tensorrt_llm/_torch/attention/backends/flashinfer.py [hunk 3]
```diff
@@ -1471,6 +1513,73 @@ def _refresh_fa2_cuda_graph_plans(self) -> None:
...
+            if not plan_info[_FI_PLAN_SPLIT_KV]:
+                # No split, no partials and no merge.
+                return
+            num_rows = (plan_info[_FI_PLAN_PADDED_BATCH_SIZE] *
+                        plan_info[_FI_PLAN_CTA_TILE_Q] * num_qo_heads)
+            v_offset = plan_info[_FI_PLAN_V_OFFSET]
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention/backends/flashinfer.py]: What if FA2 uses different layout in the future version? Can we extract the layout info from flashinfer?

File: tensorrt_llm/_torch/attention/backends/flashinfer.py (hunk 2 above)
Comment (brnguyen2) [reply tensorrt_llm/_torch/attention/backends/flashinfer.py]: Good point — dropped the length-based dispatch. The caller (the paged-prefill, ragged-prefill and decode run paths) always knows which wrapper it holds, so `zero_planned_split_kv_scratch` now takes an explicit `is_decode` and selects the plan-info layout from that rather than from the vector length. The length is only used afterward to confirm the wrapper is actually on the FA2 (prefill) or batch-decode plan these offsets describe — any other backend has a different length and is left alone — and an out-of-range extent now raises instead of touching the wrong bytes. The whole per-pass reset is also now behind an off-by-default env var (`TLLM_FLASHINFER_RESET_SPLIT_KV_SCRATCH`), so this path only runs when explicitly enabled.

File: tensorrt_llm/_torch/attention/backends/flashinfer.py (hunk 3 above)
Comment (brnguyen2) [reply tensorrt_llm/_torch/attention/backends/flashinfer.py]: FlashInfer only surfaces the plan info as a flat int vector — there's no named-field accessor in the pinned version — so these offsets are inherently coupled to that version's `PrefillPlanInfo`/`DecodePlanInfo::ToVector` order. Rather than hide that coupling, the offsets are now pinned to a known version set (`_FI_SPLIT_KV_SUPPORTED_VERSIONS`), and when the reset is enabled a FlashInfer version outside that set fails loudly at metadata build time instead of running against a possibly-reordered layout. Combined with the off-by-default gate, a future FlashInfer bump can't silently zero the wrong bytes. If this ever needs to be always-on, the right long-term fix is a named plan-info accessor upstream in FlashInfer.

File: tensorrt_llm/_torch/attention/backends/flashinfer.py [hunk 4]
```diff
@@ -254,6 +254,80 @@ class FlashInferWrappers:
...
+def _assert_split_kv_reset_supported() -> None:
+    """Reject an unpinned FlashInfer when the opt-in per-pass reset is on.
+
+    The split-KV scratch offsets are pinned to
+    ``_FI_SPLIT_KV_SUPPORTED_VERSIONS``; on any other FlashInfer version the
+    ``PrefillPlanInfo``/``DecodePlanInfo`` layout may have moved, so fail loudly
+    instead of touching the wrong workspace bytes.
+    """
+    version = getattr(flashinfer, "__version__", None)
+    if version not in _FI_SPLIT_KV_SUPPORTED_VERSIONS:
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention/backends/flashinfer.py:321]: Use a specific flashinfer here means this fix will be out-of-date in two weeks since flashinfer releases a new version every two weeks. It is hard to convince me that we need to maintain a fix for potential issue with a long term effort. Any better idea?

File: tensorrt_llm/_torch/attention/backends/flashinfer.py [hunk 5]
```diff
@@ -988,6 +1062,17 @@ def _post_init_with_buffers(self, buffers) -> None:
...
+            # split-KV partial outputs and log-sum-exps out of it and merges
+            # them afterwards, and it is never cleared between plans. Zero it
+            # once at allocation so the merged attention result cannot depend on
+            # whatever the allocation happened to land on. This is cheap (one
+            # memset per engine) and unconditional; the optional per-pass reset
+            # in zero_planned_split_kv_scratch() extends the same guarantee to
+            # later passes and is off by default.
+            self.workspace_buffer.zero_()
+            if _FI_SPLIT_KV_RESET_ENABLED:
+                _assert_split_kv_reset_supported()
```
[the remark this answers is not in the document]
Comment (brnguyen2) [reply tensorrt_llm/_torch/attention/backends/flashinfer.py:1075]: Added in 0c005f8dbe: `test_post_init_zeroes_workspace_and_checks_version` drives the real `_post_init_with_buffers()` with a stub metadata and a dirty host workspace, asserts the workspace is zeroed, and asserts `_assert_split_kv_reset_supported()` is invoked when the reset is enabled (cut off at the first CUDA allocation that follows, so it stays CPU-only).

File: tensorrt_llm/_torch/attention/backends/flashinfer.py [hunk 6]
```diff
@@ -1500,6 +1585,93 @@ def _refresh_fa2_cuda_graph_plans(self) -> None:
...
+        # fail loudly rather than zeroing memory outside the plan's two ranges.
+        total = self.workspace_buffer.numel()
+        if (v_offset < 0 or s_offset < 0 or v_offset + v_bytes > total
+                or s_offset + s_bytes > total):
+            raise RuntimeError(
+                "FlashInfer split-KV scratch extent is out of range "
+                f"(is_decode={is_decode}): tmp_v=[{v_offset}, "
+                f"{v_offset + v_bytes}), tmp_s=[{s_offset}, {s_offset + s_bytes}"
+                f"), workspace={total} bytes. The plan-info offsets are likely "
+                "stale for this flashinfer-python version.")
```
[the remark this answers is not in the document]
Comment (brnguyen2) [reply tensorrt_llm/_torch/attention/backends/flashinfer.py:1665]: Added in 0c005f8dbe: `test_out_of_range_extent_raises_and_leaves_workspace_untouched` sizes the workspace one byte short of the planned `tmp_s` end for both the prefill and decode layouts, asserts `RuntimeError`, and asserts every byte still holds the fill pattern.

File: tests/unittest/_torch/attention/test_flashinfer_attention.py [hunk 7]
```diff
@@ -1412,3 +1412,124 @@ def test_no_generations_clears_the_recorded_page_tuple():
...
+        plan_info[flashinfer_backend._FI_PLAN_SPLIT_KV] = int(split_kv)
+        return plan_info
+
+    def _decode_plan_info(self, split_kv: bool = True) -> List[int]:
+        plan_info = [0] * flashinfer_backend._FI_DECODE_PLAN_INFO_LEN
+        plan_info[flashinfer_backend.
+                  _FI_DECODE_PLAN_PADDED_BATCH_SIZE] = self.PADDED_BATCH_SIZE
+        plan_info[flashinfer_backend._FI_DECODE_PLAN_V_OFFSET] = self.V_OFFSET
+        plan_info[flashinfer_backend._FI_DECODE_PLAN_S_OFFSET] = self.S_OFFSET
+        plan_info[flashinfer_backend._FI_DECODE_PLAN_SPLIT_KV] = int(split_kv)
```
[the remark this answers is not in the document]
Comment (brnguyen2) [reply tests/unittest/_torch/attention/test_flashinfer_attention.py:1449]: Added in 0c005f8dbe: `test_plan_info_layout_matches_flashinfer_0_6_18` pins the layout constants to the literal FlashInfer 0.6.18 values — prefill (15; 0, 3, 10, 11, 14) and decode (10; 0, 1, 2, 9) — so an edited constant can no longer drift the fixtures and the implementation together.

File: tensorrt_llm/_torch/attention/backends/flashinfer.py (hunk 4 above)
Comment (brnguyen2) [reply tensorrt_llm/_torch/attention/backends/flashinfer.py:343]: Agreed with @yuxianq offline: the version pin is a stopgap and not worth the maintenance cost given flashinfer's release cadence. Plan is to upstream a FlashInfer change that exposes the split-KV plan layout directly, then query that layout for the zeroing instead of pinning `_FI_SPLIT_KV_SUPPORTED_VERSIONS`. Moving this PR back to draft until that lands upstream and we pick up the release; keeping only the unconditional allocation-time zero in the interim.

File: tensorrt_llm/_torch/attention/backends/flashinfer.py (hunk 3 above)
Comment (brnguyen2) [reply tensorrt_llm/_torch/attention/backends/flashinfer.py]: Same plan as the line-343 thread: rather than assuming the FA2 layout/length here, we'll query it from FlashInfer once the upstream plan-layout accessor lands. Draft until then.

File: tensorrt_llm/_torch/attention/backends/flashinfer.py (hunk 2 above)
Comment (brnguyen2) [reply tensorrt_llm/_torch/attention/backends/flashinfer.py]: Covered by the same follow-up as the line-343/1560 threads: the plan-info length assumption goes away once we read the layout from FlashInfer directly. Draft until the upstream accessor is available.

**Comment by brnguyen2** [comment]
Moving to draft. Per offline discussion with @yuxianq, the `_FI_SPLIT_KV_SUPPORTED_VERSIONS` version pin is a stopgap. Follow-up plan: upstream a FlashInfer change that exposes the split-KV plan layout, then query that layout for the scratch zeroing instead of pinning versions/assuming the FA2 length. This PR stays in draft until that lands upstream and we pick up the release. The bandit fix and allocation-time zeroing here remain valid regardless.
